server/lib/models/matchmaking-times.ts

```diff
@@ -0,0 +1,132 @@
+import db from '../db'
+import sql from 'sql-template-strings'
+import { MatchmakingType } from '../../../common/matchmaking'
+
+export interface MatchmakingTime {
+  id: string
+  type: MatchmakingType
+  startDate: number
+  enabled: boolean
+}
+
+function convertFromDb(props: {
+  /* eslint-disable camelcase */
+  id: string
+  matchmaking_type: MatchmakingType
+  start_date: number
+  enabled: boolean
+  /* eslint-enable camelcase */
+}): MatchmakingTime {
+  return {
+    id: props.id,
+    type: props.matchmaking_type,
+    startDate: +props.start_date,
+    enabled: props.enabled,
+  }
+}
+
+/**
+ * Retrieves a history of matchmaking times for a particular matchmaking type. History size can be
+ * provided which will be used to retrieve (at most) that amount of history on both sides of the
+ * date nearest to the current.
+ */
+export async function getMatchmakingTimesHistory(
```

Comment:
Actually, this query fetches the current applicable row, 10 future rows _and_ 10 past rows. That's mainly the reason for the complexity behind this query. I was going by the [spec](https://github.com/ShieldBattery/ShieldBattery/issues/336), which says:
- Admin UI could display next N future changes + M previous changes

Granted, N = M in this implementation :p
But yeah, I was thinking about this quite a bit (should've probably asked for clarification though), and thought it was fine to not have pagination in this case. Especially since this is only used to display a sort of snapshot of matchmaking times for admins when they're adding new times. In my next PR, which implements the user-facing things, I actually wrote a new query which fetches the future N rows, but also filters the ones that have the same `enabled` status as the one before them.

Anyway, I can split this query into multiple ones if that's what we want to do, but implementing a pagination for a list which shows current, future _and_ past rows might be a little bit tricky ^^. Unless we just stop trying to center around the current row and just get all the rows, in order and paginated. This is what we do in matchmaking map pools history btw. But there it's less likely we'll have a lot of entries in the future, unlike here, where matchmaking times might be entered by an admin well into the future.

Reply:
When I say stuff like that, I don't intend it to mean "this must be implemented as a single query on the backend." The fact that the UI is displaying that stuff just means that it must be query-able, not that it needs to be query-able as a single unit (and making it a single unit makes it really hard to do things like pagination, as you have discovered). I don't think displaying *only* N future and M previous changes is really tenable, there are going to be times where an admin needs to look up e.g. some far-past matchmaking time in order to check when something happened, or we have a lot of future times we need to look through, etc.

I would instead separate all these queries out: 
1) get the current time
2) get N times before time X
3) get N times after time X

The HTTP API function could call those 3 functions in that way (and parallelize 2 and 3), and 2 and 3 would be available for separate API calls to handle pagination stuff.

Reply:
Yeah, that sounds good. I just have a quick question. How should we display the history now in the admin panel? Just display the current time separately (somewhere on top) and then two tables for past/future times, each with their own paging?

Reply:
I'd probably make a table that starts with some rows on each side of the current time, maybe highlights the current time in some way, and has buttons or something to retrieve more rows on either side.